src/Symfony/Component/HttpClient/CurlHttpClient.php

```diff
@@ -0,0 +1,323 @@
+ *
+ * This provides fully concurrent HTTP requests, with transparent
+ * HTTP/2 push when a curl version that supports it is installed.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+final class CurlHttpClient implements HttpClientInterface
+{
+    use HttpClientTrait;
+
+    private $defaultOptions = self::OPTIONS_DEFAULTS;
+    private $noProxy;
+    private $multi;
+
+    /**
+     * @param array       $defaultOptions     Default requests' options
+     * @param int         $maxHostConnections The maximum number of connections to a single host
+     * @param string|null $noProxy            A comma separated list of hosts that do not require a proxy to be reached, even if one is specified
+     */
+    public function __construct(array $defaultOptions = [], int $maxHostConnections = 6, string $noProxy = null)
+    {
+        if ($defaultOptions) {
+            [, $this->defaultOptions] = self::prepareRequest(null, null, $defaultOptions, self::OPTIONS_DEFAULTS);
+        }
+
+        $mh = curl_multi_init();
+
+        // Don't enable HTTP/1.1 pipelining: it forces responses to be sent in order
+        if (\defined('CURLPIPE_MULTIPLEX')) {
+            curl_multi_setopt($mh, CURLMOPT_PIPELINING, CURLPIPE_MULTIPLEX);
+        }
+        curl_multi_setopt($mh, CURLMOPT_MAX_HOST_CONNECTIONS, $maxHostConnections);
+
+        $this->noProxy = $noProxy;
+
+        // Use an internal stdClass object to share state between the client and its responses
+        $this->multi = $multi = (object) [
+            'openHandles' => [],
+            'handlesActivity' => [],
+            'handle' => $mh,
+            'pushedResponses' => [],
+            'dnsCache' => [[], [], []],
+        ];
+
+        // Skip configuring HTTP/2 push when it's unsupported or buggy, see https://bugs.php.net/76675
+        if (\PHP_VERSION_ID < 70215 || \PHP_VERSION_ID === 70300 || \PHP_VERSION_ID === 70301) {
+            return;
+        }
+
+        // HTTP/2 push crashes before curl 7.61
+        if (!\defined('CURLMOPT_PUSHFUNCTION') || 0x073d00 > ($v = curl_version())['version_number'] || !(CURL_VERSION_HTTP2 & $v['features'])) {
+            return;
+        }
+
+        // Keep a dummy "onPush" reference to work around a refcount bug in PHP
+        curl_multi_setopt($mh, CURLMOPT_PUSHFUNCTION, $multi->onPush = static function ($parent, $pushed, array $rawHeaders) use ($multi) {
+            $headers = [];
+
+            foreach ($rawHeaders as $h) {
+                if (false !== $i = strpos($h, ':', 1)) {
+                    $headers[substr($h, 0, $i)] = substr($h, 1 + $i);
+                }
+            }
+
+            if ('GET' !== $headers[':method'] || isset($headers['range'])) {
+                return CURL_PUSH_DENY;
+            }
+
+            $url = $headers[':scheme'].'://'.$headers[':authority'];
+
+            // curl doesn't validate the pushed ":authority" header, but this is a MUST in the HTTP/2 RFC
```

Comment:
https://github.com/curl/curl/issues/3577
If only your review we're informed.

Reply:
Good so this is false comment now and check can be removed

Reply:
So I should write the code against an unreleased version of libcurl? Please try to be helpful.

Reply:
Well your call if you consider this important enough. IMO nature of this bug should cause request to fail anyway, so win is very slim.

Reply:
This is a pushed response, not a request. Current versions of curl do not validate pushed responses, so we must validate them using some logic to not put users at risk.

Reply:
I agree with @nicolas-grekas here, your comment is not very helpful and bit hostile, as it lacks a proper explanation.

You do have a good point about this bug though, if the `:authority` header is not validated it's in theory possible to bypass the Certificate security, which is an important feature of HTTP/2. I'm not up-to-date with the entire spec, but if this is considered big enough of a security risk this should be skipped as an unsupported version.

Yes an option might be a possible also, however experience learns that this only leads to more problems along the way. "Doesn't work? Disable peer-verify (instead of setting-up CA properly), problem solved 🤷‍♂️ - Great you just opened a big security hole instead of fixing the actual problem."

Reply:
Is that what's all the fuss about, that it lacks more convenient explanation? Explanation is on the very next line of code, I thought it's clear enough

```
// doing full validation requires evaluating cert's alt names, origin frames and even the alt-svc
// let's keep it simple and do URL-based validation for now
```